presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -43,14 +56,33 @@
      *
      * @throws NullPointerException if the element is null
      */
-    public void add(T element)
+    public Optional<CompletableFuture<?>> add(T element)
     {
         requireNonNull(element, "element is null");
-        if (finished.get()) {
-            return;
-        }
+        checkState(!finished.get());
+
         queue.add(element);
-        notEmptyFuture.get().complete(null);
+
+        int newQueueSize = queueSize.incrementAndGet();
+
+        notEmptyFutureReference.get().complete(null);
+
+        // if the queue is empty and the current future is finished, create a new one so
+        // a new readers can be notified when the queue has elements to read
+        if (newQueueSize >= maxQueueSize) {
+            CompletableFuture<?> future = notFullFutureReference.get();
+            if (future.isDone()) {
+                notFullFutureReference.compareAndSet(future, new CompletableFuture<>());
+            }
+        }
+
+        // if someone removed an element while we were updating the state, complete the current future
+        if (queueSize.get() < maxQueueSize) {
```

Comment:
See my note above. I'm pretty sure it's expected that `add` will be called after `finish`